src/docker.rs

```diff
@@ -0,0 +1,167 @@
+}
+
+#[async_trait]
+impl image::Image for DockerImage {
+    // pull ensures that the image is present locally and, if it is isn't
+    // will do the work necessary to pull it.
+    async fn pull(&self, username: String, password: String, force: bool) -> Result<()> {
+        if self.is_present_locally().await && !force {
+            debug!("✅ Skipping the pull process as the image was found locally");
+            return Ok(());
+        }
+
+        let auth = RegistryAuth::builder()
+            .username(username)
+            .password(password)
+            .build();
+
+        let pull_opts = PullOpts::builder()
+            .image(&self.repo)
+            .tag(&self.tag)
+            .auth(auth)
+            .build();
+
+        let images = self.runtime.images();
+        let mut stream = images.pull(&pull_opts);
+        while let Some(pull_result) = stream.next().await {
+            match pull_result {
+                Ok(output) => {
+                    debug!("🔧 {:?}", output);
+                }
+                Err(e) => {
+                    return Err(anyhow!("{}", e));
+                }
+            }
+        }
+
+        debug!("✅ Successfully pulled the image");
+        Ok(())
+    }
+
+    // copy_files uses the image_structs values to copy files from the
+    // image's file systems appropriately.
+    async fn copy_files(
+        &self,
+        content_path: String,
+        download_path: String,
+        write_to_stdout: bool,
+    ) -> Result<()> {
+        // Create the container
+        let container_id = match self.start().await {
+            Ok(id) => id,
+            Err(e) => {
+                return Err(anyhow!("failed to start the image: {}", e));
+            }
+        };
+
+        let mut content_path_buffer = PathBuf::new();
+        content_path_buffer.push(&content_path);
+
+        let mut download_path_buffer = PathBuf::new();
+        download_path_buffer.push(&download_path);
+
+        // Get the files from the container
+        let bytes = self
+            .runtime
+            .containers()
+            .get(&*container_id)
+            .copy_from(&content_path_buffer)
+            .try_concat()
+            .await?;
+
+        // Fail out if the buffer data processed is empty
+        if bytes.is_empty() {
+            return Err(anyhow!("failed to retrieve the files from the container"));
+        }
+
+        // Unpack the archive
+        let mut archive = Archive::new(&bytes[..]);
+        if write_to_stdout {
+            unimplemented!()
+        } else {
+            archive.unpack(&download_path_buffer)?;
+        }
+
+        info!(
+            "✅ Copied content to {} successfully",
+            download_path_buffer.display()
+        );
+
+        // Stop the container
+        match self.stop(container_id).await {
+            Ok(_) => {}
+            Err(e) => {
+                return Err(anyhow!("failed to stop the image: {}", e));
+            }
+        }
```

Comment:
nit: I think this should go under the `stop()` function for clarity

Reply:
Which part, creating the error?

Reply:
No, I meant stopping the container. Stopping and deleting a container are related, so they can both go under `stop()`

Reply:
Ah so are you saying that instead of `copy_files()` starting and stopping the container via `start()` and `stop()` we should have those functions be called outside of `copy_files()` directly in the `lib.rs` run function? Something like this:

```rust
// src/lib.rs - fn run()

// Create the container
let container_id = match container.start().await {
    Ok(id) => id,
    Err(e) => {
        return Err(anyhow!("failed to start the image: {}", e));
}

// Copy files from the image
match container
    .copy_files(
        config.content_path,
        config.download_path,
        config.write_to_stdout,
        container_id,
    )
    .await
{
    Ok(_) => {}
    Err(e) => {
        return Err(anyhow!("❌ error copying the image's files: {}", e));
    }
}

// Stop the container
match container.stop(container_id).await {
    Ok(_) => {}
    Err(e) => {
        return Err(anyhow!("failed to stop the image: {}", e));
    }
}

```